core/src/main/scala/akka/persistence/r2dbc/internal/postgres/PostgresJournalDao.scala

```diff
@@ -63,82 +63,77 @@ private[r2dbc] object PostgresJournalDao {
  * Class for doing db interaction outside of an actor to avoid mistakes in future callbacks
  */
 @InternalApi
-private[r2dbc] class PostgresJournalDao(journalSettings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+private[r2dbc] class PostgresJournalDao(journalSettings: R2dbcSettings, executorProvider: R2dbcExecutorProvider)(
+    implicit
     ec: ExecutionContext,
     system: ActorSystem[_])
     extends JournalDao {
   import JournalDao.SerializedJournalRow
   protected def log: Logger = PostgresJournalDao.log
 
-  private val persistenceExt = Persistence(system)
-
-  protected val r2dbcExecutor =
-    new R2dbcExecutor(
-      connectionFactory,
-      log,
-      journalSettings.logDbCallsExceeding,
-      journalSettings.connectionFactorySettings.poolSettings.closeCallsExceeding)(ec, system)
+  protected val persistenceExt: Persistence = Persistence(system)
 
-  protected val journalTable: String = journalSettings.journalTableWithSchema
+  protected def journalTable(slice: Int): String = journalSettings.journalTableWithSchema(slice)
   protected implicit val journalPayloadCodec: PayloadCodec = journalSettings.journalPayloadCodec
   protected implicit val tagsCodec: TagsCodec = journalSettings.tagsCodec
   protected implicit val timestampCodec: TimestampCodec = journalSettings.timestampCodec
   protected implicit val queryAdapter: QueryAdapter = journalSettings.queryAdapter
 
-  protected val (insertEventWithParameterTimestampSql, insertEventWithTransactionTimestampSql) = {
-    val baseSql =
-      s"INSERT INTO $journalTable " +
-      "(slice, entity_type, persistence_id, seq_nr, writer, adapter_manifest, event_ser_id, event_ser_manifest, event_payload, tags, meta_ser_id, meta_ser_manifest, meta_payload, db_timestamp) " +
-      "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, "
-
-    // The subselect of the db_timestamp of previous seqNr for same pid is to ensure that db_timestamp is
-    // always increasing for a pid (time not going backwards).
-    // TODO we could skip the subselect when inserting seqNr 1 as a possible optimization
-    def timestampSubSelect =
-      s"(SELECT db_timestamp + '1 microsecond'::interval FROM $journalTable " +
-      "WHERE persistence_id = ? AND seq_nr = ?)"
-
-    val insertEventWithParameterTimestampSql = {
-      if (journalSettings.dbTimestampMonotonicIncreasing)
-        sql"$baseSql ?) RETURNING db_timestamp"
-      else
-        sql"$baseSql GREATEST(?, $timestampSubSelect)) RETURNING db_timestamp"
-    }
+  protected def insertEventWithParameterTimestampSql(slice: Int): String = {
```

Comment:
pretty much all of the sql must be created for a given slice, which maps to a data partition, to resolve the table name with the data partition suffix

Might be a performance degradation with all string concatenation that has to be done for each request. Not sure if it matters? Maybe we could add some caching?